Transaction reads should not interfere with add/update/delete

andreachild opened the pull request:
https://issues.apache.org/jira/browse/TINKERPOP-3142

Changed `TinkerElementContainer` to no longer consider element reads as being 'used in a transaction' which should be reserved for element add/update/delete only. This is appropriate for `TinkerTransactionGraph` because the isolation level is 'read committed'. With this isolation level, we do not need to protect against 'unrepeatable reads' or 'phantom reads'.

Prior to this change, a read-only thread could cause a 'Conflict: element modified in another transaction' error in a separate thread which was attempting add/drop/update.

andreachild commented:
@vkagamlyk would appreciate a review of this fix, thanks!

vkagamlyk commented on tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java:
```diff
@@ -1705,6 +1706,92 @@ public void shouldPreventDropUpdateVertexInConcurrentTransactions() {
+        CountDownLatch signal = new CountDownLatch(1);
+        Thread reader = new Thread(() -> {
+            while (signal.getCount() > 0) {
+                g.V().id().toList();
```
good to have assert that result is empty or contains 1 id vertexId

andreachild replied:
Added verification to check expected vertex ids (empty or single value of vertexId), same for the edge test.

andreachild commented on tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java:
```diff
@@ -1705,6 +1706,92 @@ public void shouldPreventDropUpdateVertexInConcurrentTransactions() {
+
+        // reader thread which is constantly reading the vertex ids in a separate transaction
+        CountDownLatch signal = new CountDownLatch(1);
+        Thread reader = new Thread(() -> {
```
Increased reader threads to 2

xiazcy commented:
VOTE +1 pending CHANGELOG entry, and ideally some updates to reference doc to emphasize that non-repeatable reads and phantom reads are not protected at our isolation level.

andreachild commented:
> VOTE +1 pending CHANGELOG entry, and ideally some updates to reference doc to emphasize that non-repeatable reads and phantom reads are not protected at our isolation level.

Thanks, added changelog and updated semantics docs.

Cole-Greer commented:
VOTE +1. In my opinion the [documentation](https://tinkerpop.apache.org/docs/3.7.3/reference/#_transaction_semantics) has always been explicit regarding what isolation level we guarantee. I'm not opposed to adding an extra sentence or 2 stating that non-repeatable reads and phantom reads are not covered, but I don't believe it is necessary.
